From: Jan kara <jack@suse.cz>
Subject: btrfs: Fix busyloop in transaction_kthread()
References: bnc#730103
Patch-mainline: Never (introduced by patches.suse/btrfs-handle-EIO)

When a filesystem got aborted due do error, transaction_kthread() will
busyloop.  Fix it by going to sleep in that case as well. Maybe we should
just stop transaction_kthread() when filesystem is aborted but that would be
more complex.

Signed-off-by: Jan Kara <jack@suse.cz>

diff -rupX /crypted/home/jack/.kerndiffexclude linux-3.0-SLE11-SP2/fs/btrfs/disk-io.c linux-3.0-SLE11-SP2-1-btrfs-busyloop/fs/btrfs/disk-io.c
--- linux-3.0-SLE11-SP2/fs/btrfs/disk-io.c	2012-01-23 20:10:07.771255912 +0100
+++ linux-3.0-SLE11-SP2-1-btrfs-busyloop/fs/btrfs/disk-io.c	2012-01-25 14:06:51.447662005 +0100
@@ -1641,8 +1641,10 @@ static int transaction_kthread(void *arg
 	u64 transid;
 	unsigned long now;
 	unsigned long delay;
+	bool cannot_commit;
 
 	do {
+		cannot_commit = false;
 		delay = HZ * 30;
 		vfs_check_frozen(root->fs_info->sb, SB_FREEZE_WRITE);
 		mutex_lock(&root->fs_info->transaction_kthread_mutex);
@@ -1666,8 +1668,10 @@ static int transaction_kthread(void *arg
 
 		/* If the file system is aborted, this will always fail. */
 		trans = btrfs_join_transaction(root);
-		if (IS_ERR(trans))
+		if (IS_ERR(trans)) {
+			cannot_commit = true;
 			goto sleep;
+		}
 		if (transid == trans->transid) {
 			btrfs_commit_transaction(trans, root);
 		} else {
@@ -1682,7 +1686,8 @@ sleep:
 		} else {
 			set_current_state(TASK_INTERRUPTIBLE);
 			if (!kthread_should_stop() &&
-			    !btrfs_transaction_blocked(root->fs_info))
+			    (!btrfs_transaction_blocked(root->fs_info) ||
+			     cannot_commit))
 				schedule_timeout(delay);
 			__set_current_state(TASK_RUNNING);
 		}
